Userland/Libraries/LibWeb/Loader/FrameLoader.cpp

```diff
@@ -251,12 +251,13 @@ void FrameLoader::resource_did_load()
     if (resource()->has_encoding()) {
         dbgln("This content has MIME type '{}', encoding '{}'", resource()->mime_type(), resource()->encoding().value());
     } else {
-        dbgln("This content has MIME type '{}', encoding unknown (defaulting to 'utf-8')", resource()->mime_type());
+        dbgln("This content has MIME type '{}', encoding unknown", resource()->mime_type());
     }
 
     auto document = DOM::Document::create();
     document->set_url(url);
-    document->set_encoding(resource()->encoding().value_or("utf-8"));
+    if (resource()->has_encoding())
+        document->set_encoding(resource()->encoding().value());
```

Comment:
This relies on the document having no encoding before trying to set it, otherwise it would "keep" the previous encoding. This wouldn't be an issue with the current architecture, but could become one later if the document creation changes somehow : this limitation could go unnoticed during a refactor and cause a bug where the previous document encoding is kept on new documents.

How about changing the `set_encoding` method to `document.set_encoding(Optional<String> const&)`, and passing the optional resource encoding directly ?